apps/system/js/advanced_telemetry.js

```diff
@@ -0,0 +1,426 @@
+ *  by the HUD engine in gecko.
+ *  2) Custom App Metrics.  Apps can define their custom metrics and have them
+ *  collected in a histogram as well.  This API is defined via the
+ *  AdvancedTelemetryHelper.
+ *
+ * Metrics are gathered in gecko and requested from AdvancedTelemetry Module
+ * when it is time to ship the metrics to the server depending on the send
+ * interval desired.
+ */
+
+/* global asyncStorage, SettingsListener, uuid */
+(function(exports) {
+  'use strict';
+
+  // This is the asyncStorage key we use to persist the histograms in case of a
+  // shutdown.
+  const BATCH_KEY = 'metrics.advanced_telemetry.data.v4';
+
+  // This is the asyncStorage key we use to persist our device ID
+  const DEVICE_ID_KEY = 'metrics.advanced_telemetry.deviceID.v4';
+
+  const REASON = 'advancedtelemetry';
+
+  // Constants for events we register/unreg for
+  const IDLE = 'idle';
+  const ACTIVE = 'active';
+  const HUDMETRICS = 'advanced-telemetry-update';
+
+
+  // This is the list of event types we register handlers for
+  const EVENT_TYPES = [
+    IDLE,
+    ACTIVE,
+    HUDMETRICS
+  ];
+
+
+  // This AdvancedTelemetry() constructor is the value we export from
+  // this module. This constructor does no initialization itself: that
+  // is all done in the start() instance method. See bootstrap.js for
+  // the code that actually calls the constructor and the start method.
+  function AdvancedTelemetry() {}
+
+  // We use the acronym AT internally to save typing and make
+  // references to class variables and constants more legible.
+  const AT = AdvancedTelemetry;
+
+  AT.BatchTiming = BatchTiming;
+
+  /*
+   * Exported and configurable properties. Properties of AT are visible
+   * outside of the module and are exposed for unit testing and to enable
+   * external configuration.
+   */
+
+  // Set to true to to enable debug output.  Not recommended to turn this
+  // on except while doing deep debugging.
+  AT.DEBUG = true;
+
+  // LOGINFO method allows for debugging of initialization events and
+  // packet sending events and is at a higher level then DEBUG.
+  AT.LOGINFO = true;
+
+  // This logging function is the only thing that is not exposed through
+  // the AdvancedTelemetry contstructor or its instance.
+  function debug(...args) {
+    if (AT.DEBUG) {
+      args.unshift('[AdvancedTelemetry]');
+      console.log.apply(console, args);
+    }
+  }
+
+  // This info level is more of a information level.  It's recommended to
+  // turn this on just to see general events related to initializations,
+  // and sending of payloads, etc.
+  function loginfo(...args) {
+    if(AT.LOGINFO) {
+      args.unshift('[AdvancedTelemetryInfo]');
+      console.log.apply(console, args);
+    }
+  }
+
+  // Debugging function to handle the case where the output length exceeds that
+  // which the adb console can handle.
+  function debuglongLine(longLine) {
+    if (AT.DEBUG) {
+      var array = longLine.match(/.{1,1000}/g);
```

Comment:
not sure if you still plan to combine the `debug` and `debuglongLine` functions, but just calling out that it's still here :)